Add ERC: Hybrid Fungible Token (#244)
ethereum/ERCs: ERCS/erc-7616.md

---
eip: 7616
title: Hybrid Fungible Token
description: Defines a specification where ERC-20 compatible tokens could also be non-fungible.
author: Bryson Yip (@crypto0xbryson)
discussions-to: https://ethereum-magicians.org/t/erc-7616-the-hybrid-fungible-token/18545
status: Draft
type: Standards Track
category: ERC
created: 2024-01-07

> requires: 20

Reviewer: ```suggestion
requires: 20, 721
```

Required to understand the Specification metadata section.

    function onERC7616Received(address _operator, uint256 _fromTokenId, uint256 _toTokenId, uint256 _value, bytes calldata _data) external returns (bytes4);

}
```

#### Rules

**_approving rules:_**

This EIP provides four kinds of approving functions indicating different levels of approvals, which can be described as full level approval, token ID level approval as well as value level approval.

- `setApprovalForAll`, compatible with ERC-721 and ERC-20, SHOULD indicate the full level of approval, which means that the authorized operators are capable of managing all the ERC20 and ERC721 tokens, including the ERC721 token values, owned by the owner.

- The token ID level `approve` function, compatible with ERC-721, SHOULD indicate that the authorized operator is capable of managing only the specified token ID, including its value, owned by the owner.

- The value level `approve` function, SHOULD indicate that the authorized operator is capable of managing the specified maximum value of the specified token owned by the owner.

- For any approving function, the caller MUST be the owner or has been approved with a higher level of authority.

### Metadata

#### Metadata Extensions

> ERC-7616 metadata extensions are compatible ERC-721 metadata extensions. You can see more details in [ERC-721](./eip-721.md).

Reviewer: ```suggestion
This proposal's metadata extensions are compatible ERC-721 metadata extensions. You can see more details in [ERC-721](./eip-721.md).
```

## Rationale

### Design decision: Value transfer from token to address

The 'value' of a token is a property of the token and is not linked to an address, so to transfer the value to an address would be actually transferring it to a token owned by that address, not the address itself.

> From the implementation perspective, the process of transferring values from token to address could be done as follows: (1) create a new token for the recipient's address, (2) transfer the value to the new token from the 'source token'. So that this method is not fully independent from the ID-to-ID transfer method, and can be viewed as syntactic sugar that wraps the process described above.

Reviewer: ```suggestion
From the implementation perspective, the process of transferring values from token to address could be done as follows: (1) create a new token for the recipient's address, (2) transfer the value to the new token from the 'source token'. This method is not fully independent from the ID-to-ID transfer method, and can be viewed as syntactic sugar that wraps the process described above.
```

### Design decision: Relationship between different approval models

For semantic compatibility with ERC-721 as well as the flexibility of value manipulation of tokens, we decided to define the relationships between some of the levels of approval like that:

1. Approval of an id will lead to the ability to partially transfer values from this id by the approved operator; this will simplify the value approval for an id. However, the approval of total values in a token should not lead to the ability to transfer the token entity by the approved operator.
2. `setApprovalForAll` will lead to the ability to partially transfer values from any token, as well as the ability to approve partial transfer of values from any token to a third party; this will simplify the value transfer and approval of all tokens owned by an address.

## Backwards Compatibility

As mentioned in the beginning, this EIP is backward compatible with ERC-20.

## Reference Implementation

> - [ERC-3525 implementation](../assets/eip-3525/contracts/ERC3525.sol)

Reviewer: Why are you linking to someone else's EIP implementation?

## Security Considerations

The value level approval is isolated from ERC-721 approval models, so that approving value should not affect ERC-721 level approvals. Implementations of this EIP must obey this principle.

> Since this EIP is ERC-721 compatible, any wallets and smart contracts that can hold and manipulate standard ERC-721 tokens will have no risks of asset loss for ERC-7616 tokens due to incompatible standards implementations.

Reviewer: ```suggestion
Since this EIP is ERC-721 compatible, any wallets and smart contracts that can hold and manipulate standard ERC-721 tokens will have no risks of asset loss for hybrid fungible tokens due to incompatible standards implementations.
```

## Copyright

Copyright and related rights waived via [CC0](../LICENSE.md).

> 
> ```
> 
> ```

Reviewer: ````suggestion
````

---
eip: 7616
title: Hybrid Fungible Token
description: ERC-20 compatible tokens that could be non-fungible.
author: Bryson Yip (@fomo-protocol)
discussions-to: https://ethereum-magicians.org/t/erc-7616-the-hybrid-fungible-token
status: Draft
type: Standards Track
category: ERC
created: 2024-01-07
requires: 20, 721
---

## Abstract

> This proposal outlines a standard for hybrid fungible tokens, integrating smart contract interfaces that are compatible with the [ERC-20](./eip-20.md) token standard. It synthesizes concepts from [ERC-721](./eip-721.md) with influences from the Semi-Fungible Token model of ERC-3525, introducing an innovative approach to manage both ERC-20 and ERC-721 within a single contract.

Reviewer: ```suggestion
This proposal outlines a standard for hybrid fungible tokens, integrating smart contract interfaces that are compatible with the [ERC-20](./eip-20.md) token standard. It synthesizes concepts from [ERC-721](./eip-721.md) with influences from the Semi-Fungible Token model of [ERC-3525](./eip-3525.md), introducing an innovative approach to manage both ERC-20 and ERC-721 within a single contract.
```

> Drawing inspiration from ERC-3525(./eip-3525.md), ERC7616ValueTransferable standard incorporates a value attribute for tokens, analogous to the balance attribute in [ERC-20](./eip-20.md) tokens, to signify the token's quantitative aspect. However, it moves away from the notion of distinct 'slots' for each [ERC-721](./eip-721.md) token, opting instead to maintain [ERC-721](./eip-721.md) tokens consistently within a single slot.

Reviewer: ```suggestion
Drawing inspiration from ERC-3525, the `IERC7616ValueTransferable` interface incorporates a value attribute for tokens, analogous to the balance attribute in ERC-20 tokens, to signify the token's quantitative aspect. However, it moves away from the notion of distinct 'slots' for each ERC-721 token, opting instead to maintain ERC-721 tokens consistently within a single slot.
```

One such example is Pandora ERC404, which uniquely allocates an ERC-721 token for each ERC-20 token held. This mechanism ensures that transferring an ERC-721 token inherently includes the associated ERC-20 token, embedding a layer of fractional ownership. However, unlike other NFT projects that boast a wide array of traits, ERC404 tokens possess limited variability, thereby enhancing their fungibility.

The burgeoning trend of inscriptions, as seen on platforms like insc.evm, further underscores the versatility of this integration. Each inscription represents an NFT that encapsulates a definite token quantity, enabling fluid token transfers among NFTs, a feature that expands the utility and interaction dynamics between tokens and NFTs.

Jelly.top introduces a novel concept by decoupling NFTs from tokens. This platform allows users to either attach value to NFTs or detach it, suggesting a model where tokens can establish a base value for NFTs while enabling collectors to preserve the uniqueness of their preferred NFT designs.

> Despite the innovative strides, a fundamental compatibility issue persists between ERC-20 and ERC-721 standards. Currently, contracts are constrained to embody either standard but not both simultaneously. Shared events and functions, such as the Transfer event and balanceOf, complicate the discernibility between ERC-20 and ERC-721 interactions, posing challenges for application interoperability.

Reviewer: ```suggestion
Despite the innovative strides, a fundamental compatibility issue persists between ERC-20 and ERC-721 standards. Currently, contracts are constrained to embody either standard but not both simultaneously. Shared events and functions, such as the `Transfer` event and `balanceOf`, complicate the discernibility between ERC-20 and ERC-721 interactions, posing challenges for application interoperability.
```
